components/layout_2020/flow/float.rs

```diff
@@ -70,6 +71,114 @@ impl ContainingBlockPositionInfo {
     }
 }
 
+/// This data strucure is used to try to place non-floating content among float content.
+/// This is used primarily to place replaced content and independent formatting contexts
+/// next to floats, as the specifcation dictates.
+pub(crate) struct PlacementAmongFloats<'a> {
+    /// The [FloatContext] to use for this placement.
+    float_context: &'a FloatContext,
+    /// The current bands we are considering for this placement.
+    current_bands: VecDeque<FloatBand>,
+    /// The size of the object to place.
+    object_size: Vec2<Length>,
+    /// The minimum position in the block direction for the placement. Objects should not
+    /// be placed before this point.
+    ceiling: Length,
+}
+
+impl<'a> PlacementAmongFloats<'a> {
+    pub(crate) fn new(
+        float_context: &'a FloatContext,
+        ceiling: Length,
+        object_size: Vec2<Length>,
+    ) -> Self {
+        let current_bands = VecDeque::from([float_context.bands.find(ceiling).unwrap()]);
+        PlacementAmongFloats {
+            float_context,
+            current_bands,
+            object_size,
+            ceiling,
+        }
+    }
+
+    fn top_of_placement_for_current_bands(&self) -> Length {
+        self.ceiling.max(self.current_bands.front().unwrap().top)
+    }
+
+    fn current_bands_height(&self) -> Length {
+        assert!(self.current_bands.len() > 0);
+        self.current_bands.back().unwrap().top - self.top_of_placement_for_current_bands()
+    }
+
+    fn accumulate_enough_bands_for_block_size(&mut self) {
+        while self.current_bands_height() < self.object_size.block {
+            assert!(!self.current_bands.is_empty());
+            let next_band = self
+                .float_context
+                .bands
+                .find_next(self.current_bands.back().unwrap().top)
+                .unwrap();
+            self.current_bands.push_back(next_band);
+        }
+    }
+
+    fn calculate_viable_inline_space(&self) -> (Length, Length) {
+        let mut max_inline_start = self.float_context.containing_block_info.inline_start;
+        let mut min_inline_end = self.float_context.containing_block_info.inline_end;
+        assert!(!self.current_bands.is_empty());
+
+        for band in self.current_bands.iter().take(self.current_bands.len() - 1) {
```

Comment:
There is a problem here if we only have one band:

```html
<!DOCTYPE html>
<div style="display: flow-root; border: solid; width: 100px;">
    <div style="float: left; width: 50px; height: 50px; background: cyan"></div>
    <div style="display: flow-root; width: 50px; outline: solid red"></div>
</div>
```

I think we should enforce having at least 2 bands (except if the 1st one is already infinite?), or here not ignore the last band if there is a single one.

Reply:
Oh, good point. I've pushed a new commit which should handle this case.